### ISSUE TITLE
Gifs, Features, Announcments

### ISSUE THREAD

**Issue by andreyorst** [andreyorst/SimpleSnippets.vim#1]
# SimpleSnippets.vim
#### Subscribe if you interested.

SimpleSnippets overview:
![demo](https://user-images.githubusercontent.com/19470159/39535519-c4904d2a-4e3c-11e8-9e1c-9796515f1913.gif)

**Comment by andreyorst** [comment]
Adding new snippet:
![adding a snippet](https://user-images.githubusercontent.com/19470159/39096411-5862176e-4658-11e8-8b0a-1850cf6aa034.gif)

**Comment by andreyorst** [comment]
`SimpleSnippets#addFlashSnippet()` now uses only two parameters. Usage is changed to:

```vim
SimpleSnippets#addFlashSnippet("trigger", "snippet body")
```
More here: [doc/SimpleSnippets#addFlashSnippet](https://github.com/andreyorst/SimpleSnippets.vim/blob/68bc89c0f81d297552648a23e032913f0e3a7523/doc/SimpleSnippets.txt#L292)

**Comment by andreyorst** [comment]
![compmenu](https://user-images.githubusercontent.com/19470159/39534438-416411e0-4e3a-11e8-8b15-ae9d27c7f672.gif)
Support for Deoplete completion popup! Deoplete now lists snippet candidates with descriptions, gathered from file `ft.snippets.descriptions.txt` located at your snippet filetype folder. Or if no file found, or some triggers are not presented in this file, description is made from snippet's first line.

Also `:SimpleSnippetsList` function is available to list all snippets for current filetype:
![image](https://user-images.githubusercontent.com/19470159/39534702-e48119c2-4e3a-11e8-9ce0-a4166fc6f8d0.png)

**Comment by andreyorst** [comment]
## Repeaters and Mirroring
### This update is breaking core mechanics!

### Repeaters
Repeater is a kind of placeholder, that is paired with `normal` or `command` placeholders (`shell` placeholders got renamed to `command` placeholder, as it is more sutable name). It's  syntax is `$1` or `$2` and it must be paired with either `normal` or `command` placeholder like so: `${1:text} ... $1 ... `. As a repeater, it will repeat placeholder's body text across snippet's body.
```
#ifndef ${0:DEFINE}
#define $0
#endif /* ifndef $0 */
```

Will expand to:
```c
#ifndef DEFINE
#define DEFINE
#endif /* ifndef DEFINE */
```

Same mechanics will be applied for `command` placeholders:
```
Hello, my name is ${0!getent passwd "$USER" | cut -d : -f 5},
I'd just like to interject for a moment. What you'r reffering to as "Snippets"...
...
Best regards,
$0
```

Will expand to:
```
Hello, my name is Andrey Orst,
I'd just like to interject for a moment. What you're reffering to as "Snippets"...
...
Best regards,
Andrey Orst
```


### Mirroring
This update also changes mechanics of mirroring. **Mirror placeholders are now deprecated**. No more `${1|text} .. text .. text` crap. Repeaters are come in handy here: now you can use `normal` placeholders for that. When normal placeholder is bundled with `repeater` it becomes mirror placeholder. Like in other snippet managers, expept that this is still just a substitutions. Valid syntax now is:
```
class ${1:Name} {
public:
    $1();
    virtual ~$1();

private:
    ${0:data}
}
```

Which will expand to:
```cpp
class Name {
public:
    Name();
    virtual ~Name();

private:
    data
}
```

And upon jump to `Name` (first automatic jump in this case) all occurences of Name will be selected, and you will be prompted to enter new name for class.
If you want to mirror a command placeholder, consider wrapping it with normal placeholder, and then use repeaters for normal placeholder like so:
```
#ifndef ${2:${1!substitute(substitute(expand('%:t'), '.*', '\U&', 'g'), '\v\..*', '', 'g')}_H}
#define $2
${0:contents}
#endif /* $2 */
```

Which will be expanded to:
```cpp
#ifndef CURRENTFILENAME_H
#define CURRENTFILENAME_H
contents
#endif /* CURRENTFILENAME_H */
```

With `CURRENTFILENAME_H` selected across whole snippet body to rename it.

**Comment by andreyorst** [comment]
## Jump-able command placeholders

Now command placeholders (`${0!echo "$USER"}`, `${1!substitute(g:variable, '\v.*', '\U&', 'g'}`, etc.), which result is multi-line output can be jumped and mirrored. This still needs more testing, but all cases that I've tested already it works fine. Therefore these placeholders are jump-able again, and no wrapping with normal placeholders is needed (no more `${0:${1!echo "text"}}`).

**Comment by andreyorst** [comment]
# Backjumping

Now back jumps are possible. You can jump forward while changing blaceholders, and then jump backward on changed text. This is still search for a string so be careful what you type. It mostly stable, only newlines may cause problems.
Also now you can jump out of a snippet body. Previously if no placeholders were awailable jumping stopped on last one and stuck there until you manually leave snippet. This is now a behavior of backward jumping. Forward jumps will jumpout of snippet and editing process will be stopped so no jumps will be possible. This is how it is done in other editors, so I've decided to do the same here.

![backjumping](https://user-images.githubusercontent.com/19470159/40218859-b84e4c06-5a7b-11e8-8841-95ccbf45636f.gif)

## Adjustable triggers

Now you can set trigger key with three variables in your config:
```vim
let g:SimpleSnippetsExpandOrJumpTrigger = "<Tab>"
let g:SimpleSnippetsJumpBackwardTrigger = "<S-Tab>"
let g:SimpleSnippetsJumpToLastTrigger = "<S-j>"
```
These are default ones. If you're using build one mappings you can use this variables to adjust keys without rewriting mappings itselves. If you're using custom mappings, you still need those variables, because they used in command mode while editing mirrored placeholder.

**Comment by andreyorst** [comment]
## Testing system

Recently I've decided that SimpleSnippets should be tested, but I can't run all possible checks every commit miself. So I've created some scripts that automate this work for me.

If you expirience an issue please check if some test are failing for you. You will need `tmux`, `awk`, `sha256sum` (those should be by default in most distros, except for tmux, maybe), `vim` and `neovim` to run tests. If you're not developing a feature for SimpleSnippets you can skip one editor (for example if you don't want to install (neo)vim alongside with your (neo)vim) by modifying `run_sests.sh` in the `test` folder. However testing in both editors is higly appreciated.

Each test runs tmux session in background, sends user input, as it would actual user do, saves resulting file, and then terminates tmux. After that test checks sha256sum of result and compares it to sha256sum of reference file located in the same directory.

**Comment by andreyorst** [comment]
## SimpleSnippets.vim 0.0.8 released

**Whats new:**
- Inword placeholders.
  Inword placeholders are basically placeholders contained within another word, or connected to another words. For example: `${1:function_name}_test() {...}`. On jump you will be able to edit `function_name` to your needs.

**Comment by andreyorst** [comment]
## SimpleSnippets.vim v0.0.9 release

### Features:
- New function: `:SimpleSnippetsEditDescriptions`.
  Opens split whit descriptions for current filetype. Accepts filetype as optional parameter.
  ![descriptions](https://user-images.githubusercontent.com/19470159/40861798-127bc87a-65f3-11e8-863e-7eda41149696.gif)
- New placeholder: ${VISUAL}.
  Adds support of visual placeholder, which allows to select text and yank it to the snippet body automatically.
  ![demo](https://user-images.githubusercontent.com/19470159/40861835-24aa4a08-65f3-11e8-93ec-0a7367bf5d2a.gif)

For now I think I'm finished with majority of things in the plugin, it works fine, and serves me well already, so I won't implement much of new features in the near future. There will be bug (if any) fixes, and maybe optimisations. I don't think that with current implementation anything more is possible, but suggestions and feature requests are always welcome.

**Comment by andreyorst** [comment]
> For now I think I'm finished with majority of things in the plugin, it works fine, and serves me well already, so I won't implement much of new features in the near future.

Just kidding.

## [SimpleSnippets.vim v0.1.0 release](https://github.com/andreyorst/SimpleSnippets.vim/releases/tag/v0.1.0)

**Features**:
- New placeholder type: `choice`
  This placeholder lets your choose the candidate from the list:
  ![choice](https://user-images.githubusercontent.com/19470159/40890243-b9e65e98-677b-11e8-958f-a36c7da58251.gif)
  Choice placeholder supports mirroring. On successful jump this placeholder is converted to either `normal`, or `mirror` type.
- New option: `g:SimpleSnippets_split_horizontal`
  This option lets you define a layout to `SimpleSnippetsEdit` command. Now you can choose between `vertical` (default) and `horizontal` split.

# 

This really may be the last major release for now. The remaining limitations still must be fixed but with current implementation this may be difficult. I still don't know the proper way of handling nested placeholders (`${2:some ${1:text}` (yes previously I've tried to achieve this with command placeholders, and `${VISUAL}` can also be nested, but that's different story), and tabspots (`$1`), however these two were first limitations I've tried to withdrawn.

**Comment by andreyorst** [comment]
# [SimpleSnippets.vim v0.1.5 release](https://github.com/andreyorst/SimpleSnippets.vim/releases/tag/v0.1.5)

This release fixes issues with using `gdefault` option, as well as excludes backup files from snippet list. 

# 

I've really messed up with versioning, and the previous release here was 0.1.0, and current is 0.1.5, just because I've updated the previous, instead of making new releases. This time I wanted to write a little post here about future development.

I'm working on 0.2.0 version right now, which will fix most of limitations that current implementation have. Tabstops (in my case its empty placeholders), realtime mirroring, jumping on non-unique strings will be possible. On the downside I think that the lowest compatible version of Vim will increase to 8.1, because of new semi-realtime user input handling mechanism, that requires new features.

I'm not sure **when** i'm gonna release 0.2.0, maybe at the end of this month, maybe in the august, but I definitively will. 

_(who I am even talking with)_

**Comment by andreyorst** [comment]
### [SimpleSnippets.vim v0.1.5 release](https://github.com/andreyorst/SimpleSnippets.vim/releases/tag/v0.1.5)

This release fixes issues with using `gdefault` option, as well as excludes backup files from snippet list. 

# 

I've really messed up with versioning, and the previous release here was 0.1.0, and current is 0.1.5, just because I've updated the previous, instead of making new releases. This time I wanted to write a little post here about future development.

I'm working on 0.2.0 version right now, which will fix most of limitations that current implementation have. Tabstops (in my case its empty placeholders), realtime mirroring, jumping on non-unique strings will be possible. On the downside I think that the lowest compatible version of Vim will increase to 8.1, because of new semi-realtime user input handling mechanism, that requires new features.

I'm not sure **when** i'm gonna release 0.2.0, maybe at the end of this month, maybe in the august, but I definitively will. 

_(who I am even talking with)_

**Comment by andreyorst** [comment]
## [SimpleSnippets.vim v0.1.5 release](https://github.com/andreyorst/SimpleSnippets.vim/releases/tag/v0.1.5)

This release fixes issues with using `gdefault` option, as well as excludes backup files from snippet list. 

# 

I've really messed up with versioning, and the previous release here was 0.1.0, and current is 0.1.5, just because I've updated the previous, instead of making new releases. This time I wanted to write a little post here about future development.

I'm working on 0.2.0 version right now, which will fix most of limitations that current implementation have. Tabstops (in my case its empty placeholders), realtime mirroring, jumping on non-unique strings will be possible. On the downside I think that the lowest compatible version of Vim will increase to 8.1, because of new semi-realtime user input handling mechanism, that requires new features.

I'm not sure **when** i'm gonna release 0.2.0, maybe at the end of this month, maybe in the august, but I definitively will. 

_(who I am even talking with)_

**Comment by andreyorst** [comment]
## [SimpleSnippets.vim v0.1.6 release](https://github.com/andreyorst/SimpleSnippets.vim/releases/tag/v0.1.6)

This release adds new feature - importing snippets from other filetype. Now you can use new command: `:SimpleSnippetsImport filetype` to import `filetype` snippets to current filetype.
For example, you're editing markdown document, which contains code in it in some language, for example Rust, and you would like to use snippets for Rust in markdown, along with markdown snippets.

`:SimpleSnippetsImport rust` will make rust snippets available for current session. `SimpleSnippetsUnload rust` will unload imported snippets, if you no longer need them. You can't unload your main filetype snippets with `:SimpleSnippetsUnload`.

# 

Note about testing. Now test suite uses [vim-plug](https://github.com/junegunn/vim-plug), which is more popular plugin manager as far as I can see. Most plugin devs accept minimal `.vimrc`s with vim-plug as a plugin manager.

**Comment by andreyorst** [comment]
So. It's been a while. 

I've worked on SimpleSnippets.vim, because I wanted fast and comprehensive snippet support in Vim, without need for python dependencies or anything else. I'm pretty happy with this plugin, It serves me well, however inside it is hell of a garbage, to be honest. I've tried to do some refactoring with 0.2.0 version, but realized that I simply do not want to do it. Plugin works fine, yes, there are some limitations, but it is definitively not a bad plugin on it's own. I just feel tired when it comes to write viml. So I've decided to step out for a while. I need fresh air, so I've tried Kakoune in past few months, and now I'm exploring Emacs. TBH I kinda like it more than Vim, so maybe I'll never come back to writing viml again, and therefore there will be no 0.2.0 version of this snippet manager from me. If someone wants to take on SimpleSnippets.vim, I'll be happy to provide access to the repo. In current state plugin is stable, and no critical (or any) bugs were spotted by me. It's just not the best implementation.

I'm closing this issue for now, and I'll add a note, that much likely will point to this comment, so newcomers would be aware that the plugin isn't actively maintained.

So long.
